include/llvm-dialects/Dialect/OpMap.h

```diff
@@ -0,0 +1,356 @@
+/*
+ * Copyright (c) 2023 Advanced Micro Devices, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include "llvm-dialects/Dialect/OpDescription.h"
+#include "llvm-dialects/Dialect/Utils.h"
+#include "llvm/ADT/DenseMap.h"
+#include "llvm/ADT/SmallVector.h"
+#include "llvm/IR/IntrinsicInst.h"
+
+using namespace llvm;
+using namespace llvm_dialects;
+
+namespace llvm_dialects {
+
+// OpMap implements a map-like container that can store core opcodes,
+// intrinsics and dialect operations. It provides some lookup functionality for
+// these kinds of operations, OpDescriptions and functions as well as
+// instructions to simplify working with dialects in scenarios requiring
+// association of dialect operations with certain values.
+template <typename ValueT> class OpMap final {
+  using DialectOpKey = std::pair<StringRef, bool>;
+
+  struct DialectOpKV final {
+    DialectOpKey Key;
+    ValueT Value;
+
+    bool operator==(const DialectOpKV &other) const {
+      return Key.first == other.Key.first && Key.second == other.Key.second &&
+             Value == other.Value;
+    }
+
+    bool operator==(const OpDescription &desc) const {
+      bool isOverload =
+          desc.getKind() == OpDescription::Kind::DialectWithOverloads;
+      return Key.first == desc.getMnemonic() && Key.second == isOverload;
+    }
+  };
+
+public:
+  OpMap() = default;
+
+  // -------------------------------------------------------------
+  // Convenience constructor to initialize the OpMap from a set of
+  // OpDescription/Value pairs.
+  // -------------------------------------------------------------
+  OpMap(std::initializer_list<std::pair<OpDescription, ValueT>> vals) {
+    for (const std::pair<OpDescription, ValueT> &val : vals) {
+      insert_or_get(val.first, val.second);
+    }
+  }
+
+  // -------------------------------------------------------------
+  // Comparison operator overloads.
+  // -------------------------------------------------------------
+  bool operator==(const OpMap &rhs) const {
+    return m_coreOpcodes == rhs.m_coreOpcodes &&
+           m_intrinsics == rhs.m_intrinsics && m_dialectOps == rhs.m_dialectOps;
+  }
+
+  bool operator!=(const OpMap &rhs) const { return !(*this == rhs); }
```

Comment:
I don't think these `operator==/!=` implement the correct semantics, since they depend on the ordering of the dialect ops vector. Do we need them in the first place?

Reply:
Right now, unfortunately yes, otherwise the comparison tests don't work (see the `OpMapIRTests`).
You are right on the dependency of the ordering. Maybe it makes sense to always sort the vector once it was mutated.

Reply:
I just went with the approach to do a quick pairwise lookup, which should work fine.